### PR TITLE
mse_loss added 

### PR DESCRIPTION
**Pull request by ayushi2626**
Mean square loss added in ivy/ivy/functional/ivy/losses.py
Close https://github.com/unifyai/ivy/issues/16654

### CODE REVIEW

File: ivy_tests/test_ivy/test_misc/test_mse_loss.py [hunk 1]
```diff
@@ -0,0 +1,11 @@
+import ivy
+import ivy.functional.ivy.losses as l
+import numpy as np
+def mse_loss_test(x,y):
```
Comment (Sam-Armstrong) [inline ivy_tests/test_ivy/test_misc/test_mse_loss.py:4]: The test should be located in ivy_tests/test_ivy/test_functional/test_nn/test_losses.py, and follow the same test style as the tests found in there. You can find more details about testing here https://unify.ai/docs/ivy/overview/deep_dive/ivy_tests.html.

File: test_file.py [hunk 2]
```diff
@@ -0,0 +1,5 @@
+import ivy
+from ivy.functional.ivy.losses import mse_loss
+x = ivy.Array([1,1,1])
+y = ivy.Array([0,0,0])
+print(mse_loss(x,y))
```
Comment (Sam-Armstrong) [inline test_file.py:5]: File should be removed.

File: ivy/functional/ivy/losses.py [hunk 3]
```diff
@@ -381,4 +382,47 @@ def sparse_cross_entropy(
...
+    ret
+        the mean square loss between two arrays
+
+    Example
+    -------------------
+    >>x = [1,1,1]
+    >>y = [0,0,0]
+    >>z = ivy.mse_loss(x,y)
+    >>print(z)
+    ivy.array([1])
```
Comment (Sam-Armstrong) [inline ivy/functional/ivy/losses.py:415]: Please can you change '>>x' to '>>> x' for example? To keep the syntax consistent and readable :). Also, some more examples with labels would be great. Have a look at the docstring examples on binary_cross_entropy or sparse_cross_entropy. More details here: https://unify.ai/docs/ivy/overview/deep_dive/docstring_examples.html#docstring-examples.

File: ivy/functional/ivy/losses.py [hunk 4]
```diff
@@ -381,4 +382,47 @@ def sparse_cross_entropy(
     true = ivy.one_hot(true, pred.shape[axis])
     return ivy.cross_entropy(
         true, pred, axis=axis, epsilon=epsilon, reduction=reduction, out=out
-    )
+
+
+        )
+def mse_loss(true : Union[ivy.Array, ivy.NativeArray],pred : Union[ivy.Array, ivy.NativeArray],axis : int=-1,out : Optional[ivy.Array]=None)-> ivy.Array:
```
Comment (Sam-Armstrong) [inline ivy/functional/ivy/losses.py:388]:
```suggestion
@handle_exceptions
@handle_nestable
@handle_array_like_without_promotion
@inputs_to_ivy_arrays
@handle_array_function
def mse_loss(
    true: Union[ivy.Array, ivy.NativeArray],
    pred: Union[ivy.Array, ivy.NativeArray],
    /,
    *,
    axis: int = -1,
    out: Optional[ivy.Array] = None
) -> ivy.Array:
```

File: ivy/functional/ivy/losses.py [hunk 5]
```diff
@@ -381,4 +382,47 @@ def sparse_cross_entropy(
     true = ivy.one_hot(true, pred.shape[axis])
     return ivy.cross_entropy(
         true, pred, axis=axis, epsilon=epsilon, reduction=reduction, out=out
-    )
+
+
+        )
```
Comment (Sam-Armstrong) [inline ivy/functional/ivy/losses.py:387]: Please ensure that there are two lines of whitespace between each function.

File: ivy/functional/ivy/losses.py [hunk 6]
```diff
@@ -381,4 +382,47 @@ def sparse_cross_entropy(
     true = ivy.one_hot(true, pred.shape[axis])
     return ivy.cross_entropy(
         true, pred, axis=axis, epsilon=epsilon, reduction=reduction, out=out
-    )
+
```
Comment (Sam-Armstrong) [inline ivy/functional/ivy/losses.py:385]: Keep the end of this function as it was before.

File: ivy/functional/ivy/losses.py [hunk 7]
```diff
@@ -381,4 +382,47 @@ def sparse_cross_entropy(
...
+    Parameters
+    ----------
+
+    true
+        input array containing true labels
+    pred
+        input array containing predicted labels
+    axis
+        the axis along which mse is computed. By default, it is the last axis (-1)
```
Comment (Sam-Armstrong) [inline ivy/functional/ivy/losses.py:400]:
```suggestion
        the axis along which mse is computed. Default: ``-1``
```

File: ivy/functional/ivy/losses.py [hunk 8]
```diff
@@ -381,4 +382,47 @@ def sparse_cross_entropy(
...
+     
+    true = ivy.negative(true)
+    diff = ivy.add(pred,true,axis=axis)
+    s_diff = ivy.multiply(diff,diff,axis =axis)
+    s_diff = ivy.sum(s_diff,axis = axis)
+    mse_diff = s_diff/true.shape[-1]
+    return mse_diff
+
+
+
```
Comment (Sam-Armstrong) [inline ivy/functional/ivy/losses.py:426]: Remove excess whitespace. There should be one line of whitespace at the end of the file.

**Review by Sam-Armstrong — CHANGES_REQUESTED** [review]
Hi @ayushi2626, thanks for making the PR! I've left some comments, please could you take a look and try to address them? Also please could you remove the changes to the .idea files and your test_file.py? Thanks 😊

**Comment by ayushi2626** [comment]
> Hi @ayushi2626, thanks for making the PR! I've left some comments, please could you take a look and try to address them? Also please could you remove the changes to the .idea files and your test_file.py? Thanks 😊

Hi. Thank you so much for the feedback. I will make the required changes. But just one doubt. I have not made any changes to .idea.py file.

**Comment by Sam-Armstrong** [comment]
@ayushi2626 I mean .idea/ivy.iml, .idea/misc.xml, etc.
Look at https://github.com/unifyai/ivy/pull/16735/files

**Comment by ayushi2626** [comment]
Hi. I have made the changes. Should I create a new pull request?

**Comment by Sam-Armstrong** [comment]
@ayushi2626 it just says you have removed one line of whitespace at the end of the file. Have you tried adding that line back in?

**Comment by Sam-Armstrong** [comment]
@ayushi2626 I've removed a couple of unnecessary changes that you made, but unfortunately I can't remove your change to .github/workflows/test-ivy-nn.yml, as I don't have the permissions for it. But you just need to add 1 new line with nothing on it (not even any spaces) to the end of the file to remove the change.

**Review by Sam-Armstrong — CHANGES_REQUESTED** [review]
@ayushi2626 I've made a couple of updates so please `git pull` before doing anything else 😅. It looks like the your test is failing, so please can you try to get it working? You can run the test using the command `pytest ivy_tests/test_ivy/test_functional/test_nn/test_losses.py::test_mse_loss` from the terminal.

**Comment by ayushi2626** [comment]
> @ayushi2626 I've removed a couple of unnecessary changes that you made, but unfortunately I can't remove your change to .github/workflows/test-ivy-nn.yml, as I don't have the permissions for it. But you just need to add 1 new line with nothing on it (not even any spaces) to the end of the file to remove the change.

Oh okay sure. Thank you

**Comment by ayushi2626** [comment]
> @ayushi2626 I've made a couple of updates so please `git pull` before doing anything else 😅. It looks like the your test is failing, so please can you try to get it working? You can run the test using the command `pytest ivy_tests/test_ivy/test_functional/test_nn/test_losses.py::test_mse_loss` from the terminal.

Yes I will look into the same and get it working. Thank you very much

**Comment by ayushi2626** [comment]
Hello. I have been looking into the testing of the function but there is an import error that is not getting rectified. I had asked in the discord server too but the error still persists.

**Comment by ivy-leaves** [comment]
If you are working on an open task, please **edit** the PR description to link to the issue you've created.

For more information, please check [ToDo List Issues Guide](https://unify.ai/docs/ivy/overview/contributing/the_basics.html#todo-list-issues).

Thank you :hugs:

**Comment by Sam-Armstrong** [comment]
@ayushi2626 I'd recommend running the tests with this command to get a clearer view of the stack trace:
pytest -p no:warnings --tb=short ivy_tests/test_ivy/test_functional/test_nn/test_losses.py::test_mse_loss

From this I can immediately see one problem 'negative() got an unexpected keyword argument 'axis'' - you are passing a axis kwarg to the negative function which has no axis kwarg.

**Comment by ayushi2626** [comment]
I am using docker and the test is now been executed. I have removed the syntactical error in the code but I have been getting an error: ivy.utils.exceptions.IvyBackendException and KeyError: 'cpu' Can you just explain what it means? Have I not imported any library or is it something else? Thank you very much

**Comment by Sam-Armstrong** [comment]
@ayushi2626 are you getting this for all the test cases, or just for one of the backend frameworks?

**Comment by ayushi2626** [comment]
I am getting KeyError:'cpu" for numpy backend. And for the rest of the backend, I am getting the ivy.utils.exceptions.IvyBackendException error

**Comment by Sam-Armstrong** [comment]
I think you can debug and fix the tests for the other backends then right? Ignore the numpy backend for now if you're getting KeyError

**Comment by ayushi2626** [comment]
Okay got it. I am working on it. If I have further questions I will let you know. Thank you very much

**Comment by ayushi2626** [comment]
Hello. I was debugging the code and was facing this one issue. There is an AssertionError, and it says that the value returned by the function isn't the same as the value of the out array. I don't seem to understand why this is happening. I printed the out and return values both of which are the same but the error still persists. Could you pls help me with this?

**Comment by Sam-Armstrong** [comment]
@ayushi2626 Can you push your latest changes? I'm still getting the error "negative() got an unexpected keyword argument 'axis'"

**Comment by ayushi2626** [comment]
Hello. I have pushed the changes I have made.

**Comment by ayushi2626** [comment]
Hello. It says that there is an extra newline added in losses.py file after the cross_entropy function. But I just left two newlines between each function as you said. Could you please help me with that? Thanks!!

File: ivy/functional/ivy/losses.py [hunk 9]
```diff
@@ -382,3 +382,63 @@ def sparse_cross_entropy(
...
+    ivy.array([0.34])
+    """
+
+    true = ivy.negative(true)
+    diff = ivy.add(pred,true)
+    s_diff = ivy.multiply(diff,diff)
+    s_diff = ivy.sum(s_diff,axis = axis)
+    mse_diff = s_diff/true.shape[axis]
+    out = mse_diff
+    return mse_diff
```
Comment (Sam-Armstrong) [inline ivy/functional/ivy/losses.py:444]:
```suggestion
    s_diff = ivy.square(true - pred)
    return ivy.mean(s_diff, axis=axis, out=out)
```

**Review by Sam-Armstrong — CHANGES_REQUESTED** [review]
@ayushi2626 I can't see any extra line so I think we're ok 🙂 apologies for the slow review, I've been very busy. 

I've suggested a simpler implementation which fixes the problem, as we are now utilising the 'out' argument. However, there seems to be another problem arising with this change, so please could you try debugging this. Feel free to let me know if you get stuck. Thanks!

**Comment by ayushi2626** [comment]
Okay sure I will make the changes and let you know if there are any other problems. Thank you so much!!

**Comment by Sam-Armstrong** [comment]
@ayushi2626 Have you managed to make any progress on the changes? 🙂

**Comment by ayushi2626** [comment]
Hello. I am so sorry for the delay.  I had made the changes and pushed it but there were few errors. But I could not rectify that as the docker environment is not working on my system. I am working on fixing that. Can I report back the results in sometime.

**Comment by ayushi2626** [comment]
Hello. My docker and git bash have stopped working. In git bash, it says that flake8 is not compatible with the Python version I have. The last I checked was that with the changes, there was IndexError which shouldn't happen as for the test function, the values given to the axis are 0 and 1. I have been trying to debug it for days now but still, it has been the same. So sorry for the delay.

**Comment by Sam-Armstrong** [comment]
@ayushi2626 No worries 🙂 which version of python are you using?

**Comment by ayushi2626** [comment]
I think my Python version is either 3.9 or 3.10. Whenever I try to commit the changes through git bash, it says that flake8 isn't compatible with the version of Python which I am using, as flake8 works only with Python versions 3.7.5 and below.

**Comment by Sam-Armstrong** [comment]
`pip install --upgrade flake8` maybe?

**Comment by ayushi2626** [comment]
Hey, I did that too, the error still persists.

**Comment by Sam-Armstrong** [comment]
@ayushi2626 I'm not really sure what to suggest here, other than maybe try rebuilding the docker container or starting to whole setup process from scratch

**Comment by ayushi2626** [comment]
Okay, I will rebuild the entire thing and let you know.

**Comment by sushmanthreddy** [comment]
@Sam-Armstrong @ayushi2626 can i work on this, as it is inactive since so many days.. 

Pls let me knew if the issue is free

**Comment by ayushi2626** [comment]
@Sam-Armstrong I am not able to resolve the problem. I think the code I have written is fine but I am not sure why my git isn't working with the repo

**Comment by NripeshN** [comment]
Hey @ayushi2626 
Looks like your tests are failing. Could you please take a look?

**Comment by ivy-seed** [comment]
This PR has been labelled as stale because it has been inactive for more than 7 days. If you would like to continue working on this PR, then please add another comment or this PR will be closed in 7 days.

**Comment by Sarvesh-Kesharwani** [comment]
Hi @ayushi2626,
you branch it looks like is outdated,
image
please update it with latest changes from upstream.
And let me know then.

**Comment by ivy-seed** [comment]
This PR has been closed because it has been marked as stale for more than 7 days with no activity.

**Pull request closed without merging**